main.tf

```diff
@@ -122,6 +158,123 @@ resource "observe_dataset" "ingress_logs" {
+    EOF
+  }
+
+  stage {
+    alias = "response_latency"
+    input = "filtered_metrics"
+
+    pipeline = <<-EOF
+        make_col request_count:float64(case(metric="nginx_ingress_controller_response_duration_seconds_count",value, true, float64_null())),
+          request_seconds:float64(case(metric="nginx_ingress_controller_response_duration_seconds_sum",value, true, float64_null()))
+            
+        timestats request_count: any_not_null(request_count), request_seconds: any_not_null(request_seconds), group_by(clusterUid, namespace, ingress, pod, node, container, labels)
+        make_col 
+          value: float64(request_seconds / request_count),
+          metric:"nginx_ingress_controller_response_latency_per_request"
+        
+        filter not is_null(value)
+        drop_col request_count, request_seconds
+      EOF
+  }
+
+  stage {
+    alias = "upstream_latency"
+    input = "filtered_metrics"
+
+    pipeline = <<-EOF
+      make_col request_count:float64(case(metric="nginx_ingress_controller_ingress_upstream_latency_seconds_count",value, true, float64_null())),
+        request_seconds:float64(case(metric="nginx_ingress_controller_ingress_upstream_latency_seconds_sum",value, true, float64_null()))
+          
+      timestats request_count: any_not_null(request_count), request_seconds: any_not_null(request_seconds), group_by(clusterUid, namespace, ingress, pod, node, container, labels)
+      make_col 
+        value: float64(request_seconds / request_count),
+        metric:"nginx_ingress_controller_response_upstream_latency_per_request"
+      
+      filter not is_null(value)
+      drop_col request_count, request_seconds
+    EOF
+  }
+
+  stage {
+    input = "filtered_metrics"
+
+    pipeline = <<-EOF
+      union @request_latency
+      union @response_latency
+      union @upstream_latency
+      
+      make_col
+        exported_namespace:string(labels.exported_namespace),
+        controller_pod:string(labels.controller_pod),
+        controller_namespace:string(labels.controller_namespace)
+
+      interface "metric", metric:metric, value:value
+
+      ${join("\n\n", [for metric, options in local.nginx_metrics : indent(2, format("set_metric options(\n%s\n), %q", join(",\n", [for k, v in options : format("%s: %q", k, v)]), metric))])}
+    EOF
+  }
+
+}
+
+resource "observe_link" "nginx_metrics" {
+  workspace = var.workspace.oid
+  source    = var.enable_nginx_ingress_metrics ? observe_dataset.nginx_metrics[0].oid : ""
+  target    = each.value.target
+  fields    = each.value.fields
+  label     = each.key
+
+  for_each = var.enable_nginx_ingress_metrics ? merge({
+    "Ingress" = {
+      target = var.kubernetes.ingress.oid
+      fields = ["ingress:name", "exported_namespace:namespace", "clusterUid"]
+    },
+    "Pod" = {
+      target = var.kubernetes.pod.oid
+      fields = ["controller_pod:name", "controller_namespace:namespace", "clusterUid"]
+    }
+    }, var.metrics_link_targets
```

Comment:
Do you expect to link this anywhere else?

Reply:
Left that in case someone had another dataset they wanted to link the metrics to.  I can kill it if it that's not necessary.